pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,309 @@
+
+        **Examples:**
+
+        Define a Hamiltonian cost function to minimize.
+        >>> hamiltonian = qml.Hamiltonian(coeffs=[-1.]*3,
+        ...observables=[qml.PauliX(0), qml.PauliZ(1), qml.PauliY(0)@qml.PauliX(1)])
+        Create an initial state and return the expectation value of the Hamiltonian.
+        >>> @qml.qnode(qml.device("default.qubit", wires=2))
+        ... def quant_fun():
+        ...     qml.RX(0.1, wires=[0])
+        ...     qml.RY(0.5, wires=[1])
+        ...     qml.CNOT(wires=[0,1])
+        ...     qml.RY(0.6, wires=[0])
+        ...     return qml.expval(hamiltonian)
+
+        Instatiate the optimizer with the initial circuit and the cost function. Set the stepsize
+        accordingly.
+        >>> opt = LieGradientOptimizer(circuit = quant_fun, stepsize=0.1)
+        Applying 10 steps gets us close the ground state of E=-2.23
+        >>> for step in range(10):
+        ...    print(step)
+        ...    cost = opt.step_and_cost()
+        ...    print(cost)
+
+        """
+        if not isinstance(circuit, qml.QNode):
+            raise TypeError(
+                f"`circuit` must be a `qml.QNode`, " f"received {type(circuit)} "
+            )
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        if not isinstance(circuit.func().obs, qml.Hamiltonian):
+            raise TypeError(
+                f"`circuit` must return the expectation value of a `qml.Hamiltonian`,"
+                f" "
+                f"received {type(circuit.func().obs)} "
+            )
+        self.nqubits = max(circuit.device.wires) + 1
+
+        if self.nqubits > 4:
+            print(
+                "WARNING: The exact Lie gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow."
+            )
+        if restriction is not None and not isinstance(restriction, qml.Hamiltonian):
+            raise TypeError(f"`restriction` must be a `qml.Hamiltonian`, received {type(restriction)}")
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators(restriction)
+        self.exact = exact
+        self.hamiltonian = circuit.func().obs
+        self.coeffs, self.observables = self.hamiltonian.terms
+        self.stepsize = stepsize
+
+    def step_and_cost(self):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Returns:
+            Float corresponding to the energy of the the circuit.
+
+        """
+        omegas = self.get_omegas()
+        non_zero_lie_algebra_elements = []
+        non_zero_omegas = []
+        for i, element in enumerate(omegas):
+            if not np.isclose(element, 0):
+                non_zero_lie_algebra_elements.append(self.lie_algebra_basis_names[i])
+                non_zero_omegas.append(-omegas[i])
```

Comment:
I wonder if you could vectorize this for loop, using `np.nonzero` or `np.where`?

Reply:
I figured out a way to do this!